guidance/_chat.py

```diff
@@ -0,0 +1,210 @@
+        # Can possibly accomplish this with an "aliases" dictionary that maps all aliases to the canonical key in cache
+        if chat_template in CHAT_TEMPLATE_CACHE:
+            return CHAT_TEMPLATE_CACHE[chat_template]
+        # TODO: Add logic here to try to auto-create class dynamically via _template_class_from_string method
+    
+    # Only warn when a user provided a chat template that we couldn't load
+    if chat_template is not None:
+        warnings.warn(f"""Chat template {chat_template} was unable to be loaded directly into guidance.
+                        Defaulting to the ChatML format which may not be optimal for the selected model. 
+                        For best results, create and pass in a `guidance.ChatTemplate` subclass for your model.""")
+    
+    # By default, use the ChatML Template. Warnings to user will happen downstream only if they use chat roles.
+    return ChatMLTemplate
+        
+
+def _template_class_from_string(template_str):
+    """Utility method to try to create a chat template class from a string."""
+    # TODO: Try to build this, perhaps based on passing unit tests we create?
+    pass
+
+
+# CACHE IMPLEMENTATIONS:
+
+# --------------------------------------------------
+# @@@@ ChatML @@@@
+# --------------------------------------------------
+# Note that all grammarless models will default to this syntax, since we typically send chat formatted messages.
+chatml_template = "{% for message in messages %}{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
+class ChatMLTemplate(ChatTemplate):
+    template_str = chatml_template
+
+    def get_role_start(self, role_name):
+        return f"<|im_start|>{role_name}\n"
+        
+    def get_role_end(self, role_name=None):
+        return "<|im_end|>\n"
+    
+CHAT_TEMPLATE_CACHE[chatml_template] = ChatMLTemplate
+
+
+# --------------------------------------------------
+# @@@@ Llama-2 @@@@
+# --------------------------------------------------
+# [05/08/24] https://huggingface.co/meta-llama/Llama-2-7b-chat-hf/blob/main/tokenizer_config.json#L12
+llama2_template = "{% if messages[0]['role'] == 'system' %}{% set loop_messages = messages[1:] %}{% set system_message = messages[0]['content'] %}{% else %}{% set loop_messages = messages %}{% set system_message = false %}{% endif %}{% for message in loop_messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% if loop.index0 == 0 and system_message != false %}{% set content = '<<SYS>>\\n' + system_message + '\\n<</SYS>>\\n\\n' + message['content'] %}{% else %}{% set content = message['content'] %}{% endif %}{% if message['role'] == 'user' %}{{ bos_token + '[INST] ' + content.strip() + ' [/INST]' }}{% elif message['role'] == 'assistant' %}{{ ' '  + content.strip() + ' ' + eos_token }}{% endif %}{% endfor %}"
```

Comment:
Can this be written a little more comprehensibly, but have the newlines/spaces inserted for human readability stripped out on initialisation?

Reply:
I don't think I understand jinja2 well enough to know if changes I manually make for readability can compromise intrinsics of the template :|. I'd be a bit scared to do that.

Reply:
The idea would be you'd wrap the readable template in something like `textwrap.dedent()`. You could even add a test to verify that it has been converted to what it's supposed to be. `assert string_from_file == "What the string should be"`

Reply:
I think we can do a round of aesthetics updates after this PR -- we should further extend the ChatTemplate class to have a nice __repr__ for example. I'll leave off these types of changes from this code change.

Reply:
Fair enough given time constraints